api/openai/chat.go

```diff
@@ -124,31 +125,52 @@ func ChatEndpoint(cm *config.ConfigLoader, o *options.Option) func(c *fiber.Ctx)
 			}
 			r := config.Roles[role]
 			contentExists := i.Content != nil && *i.Content != ""
-			if r != "" {
-				if contentExists {
-					content = fmt.Sprint(r, " ", *i.Content)
+			if config.TemplateConfig.ChatMessage != "" {
+				chatMessageData := model.ChatMessageTemplateData{
+					SystemPrompt: config.SystemPrompt,
+					Role:         r,
+					RoleName:     role,
+					Content:      *i.Content,
+					MessageIndex: messageIndex,
 				}
-				if i.FunctionCall != nil {
-					j, err := json.Marshal(i.FunctionCall)
-					if err == nil {
-						if contentExists {
-							content += "\n" + fmt.Sprint(r, " ", string(j))
-						} else {
-							content = fmt.Sprint(r, " ", string(j))
-						}
-					}
+				templatedChatMessage, err := o.Loader.TemplateForChatMessage(config.TemplateConfig.ChatMessage, chatMessageData)
+				if err != nil {
+					log.Error().Msgf("error processing message %+v using template \"%s\": %v. Skipping!", chatMessageData, config.TemplateConfig.ChatMessage, err)
+					continue
 				}
-			} else {
-				if contentExists {
-					content = fmt.Sprint(*i.Content)
+				if templatedChatMessage == "" {
+					log.Warn().Msgf("template \"%s\" produced blank output for %+v. Skipping!", config.TemplateConfig.ChatMessage, chatMessageData)
+					continue
```

Comment:
maybe instead of skipping here would be better to skip templating and using the message as-is? making it disappear would be probably confusing

Reply:
Well - I think we'd theoretically want to allow a template to ignore a message and return "" - so the continue here really only avoids appending "" on to a string, right?

I admit it's probably not super _clear_ that's the only effect of that statement, so I'll see if I can revise it!

The Warn is because I assume rarely the user will intend for a "" to be returned from a template, so we shouldn't call that an outright _error_, but it also is worth making visible when `debug=true` since most of the time it's probably a template logic error.